rtengine/imageio.cc

```diff
@@ -1025,44 +1039,46 @@ int ImageIO::saveJXL(const Glib::ustring &fname, float quality) const
     basic_info.ysize = height;
 
     JxlPixelFormat pixel_format;
-    std::size_t size = 0;
 
     pixel_format = JxlPixelFormat{3, JXL_TYPE_FLOAT, JXL_NATIVE_ENDIAN, 0};
     basic_info.bits_per_sample = 32;
     basic_info.exponent_bits_per_sample = 8;
-    size = 3 * 4 * width * height;
-
-    JxlBitDepth bitdepth = {
-        JXL_BIT_DEPTH_FROM_PIXEL_FORMAT, basic_info.bits_per_sample,
-        basic_info.exponent_bits_per_sample
-    };
 
     JxlEncoderFrameSettings * frame_settings = JxlEncoderFrameSettingsCreate(enc.get(), nullptr);
     JxlEncoderSetFrameLossless(frame_settings, false);
 
     const float distance = JxlEncoderDistanceFromQuality(quality);
     JxlEncoderSetFrameDistance(frame_settings, distance);
+
+#if JPEGXL_NUMERIC_VERSION >= JPEGXL_COMPUTE_NUMERIC_VERSION(0, 8, 0)
+    JxlBitDepth bitdepth = {
+        JXL_BIT_DEPTH_FROM_PIXEL_FORMAT, basic_info.bits_per_sample,
+        basic_info.exponent_bits_per_sample
+    };
     JxlEncoderSetFrameBitDepth(frame_settings, &bitdepth);
+#endif
 
     if (JxlEncoderSetBasicInfo(enc.get(), &basic_info) != JXL_ENC_SUCCESS) {
         std::cerr << "Error: JxlEncoderSetBasicInfo failed" << std::endl;
         return IMIO_CANNOTWRITEFILE;
     }
 
-    // Always use sRGB for JXL.
-    // ICC doesn't work as desired/expected.
     JxlColorEncoding color_encoding = {};
     JxlColorEncodingSetToSRGB(&color_encoding, false);
 
+#if JPEGXL_NUMERIC_VERSION >= JPEGXL_COMPUTE_NUMERIC_VERSION(0, 10, 0)
+    if (!profileData.empty()) {
+        JxlEncoderSetICCProfile(enc.get(), (const uint8_t *)profileData.data(), profileData.size());
+    } else
+#endif
     if (JxlEncoderSetColorEncoding(enc.get(), &color_encoding) != JXL_ENC_SUCCESS) {
-        std::cerr << "Error: JxlEncoderSetColorEncoding failed" << std::endl;
-        return IMIO_CANNOTWRITEFILE;
+        std::cerr << "Warning: JxlEncoderSetColorEncoding failed" << std::endl;
     }
 
-    std::vector<std::uint8_t> imagebuffer(4 * 3 * width * height);
+    std::vector<std::uint8_t> imagebuffer((std::size_t) 3 * width * height * sizeof(float));
```

Comment:
No C-style casts, please.

Also, if this computation appears more than once, better create a new constant, like:

```c++
const std::size_t required_raw_line_size = sizeof(float) * 3 * width;
```

A `static_cast<std::size_t>` is not needed here, because the `sizeof` operator returns a `std::size_t` (C++11, §5.3.3 ¶6).

Reply:
CodeQL check fails without casts.

Reply:
Yeah, but at the end.

Reply:
As I understand it, `3 * width` might already overflow `int`, so `int * int * size_t` doesn't help.  `size_t * int * int` -> `size_t * int` -> `size_t` on the other hand should solve the problem.